### ISSUE TITLE
Some overseen advantages of the pairwise losses

### ISSUE THREAD

**Issue by ducha-aiki** [jeromerony/dml_cross_entropy#3]
Hi,

I really love your paper, but would like to point some problems with "unary" softmax. 

1. DML losses can work with infinite amount of classes. Pairwise CE can as well, unlike the traditional softmax. And by "infinite" one would mean either "not fitting into the memory", or "changing on the fly as new data comes". 
2. "Softmax bottleneck" While it is true that pairwise losses need to pay attention to sampling, once as you could afford big enough batch size, it is not the problem anymore. 

To conclude and if you are interested in the counter-examples - consider local descriptor learning -- it is a specific example of metric learning. Lets consider "Liberty" from PhotoTour dataset 
(available at torchvision https://pytorch.org/docs/stable/torchvision/datasets.html#phototour ). It consists of 450k samples, 161k classes. 
All the sota, or near-sota methods e.g. https://github.com/DagnyT/hardnet )  use a variant of a triplet loss with hard-negative mining. I am not aware of any successful application of the cross-entropy there.

**Comment by jeromerony** [comment]
Hi,
Thanks for you interest in our work.

Considering your first point, you could imagine many strategies to overcome the limitations you mention. If the problem is memory, a solution that I've heard is used by a (large) company is to perform the softmax on a subset of the classes, which seems to work nicely. If the problem is adding (or removing) classes, you could add a class and initialize it to minimize cross-entropy while fixing the rest. I think these are open research questions for classification.

I am more troubled by your second point. What we wanted to show is that you can obtain competitive performance by completely discarding these complex sampling strategies. Increasing batch size indefinitely has many obvious disadvantages (especially in practical applications); it is not simply a matter of "being able to afford it". It also hinders reproducibility and makes the comparisons unfair between papers. I would argue that having a fixed batch size and sampling strategy would make much more sense in the context of research.

I would be curious to try and train a model with cross-entropy on the PhotoTour dataset which indeed would need some adjustments (for memory constraints) but I would not be surprised if it worked well.

**Comment by ducha-aiki** [comment]
> If the problem is memory, a solution that I've heard is used by a (large) company is to perform the softmax on a subset of the classes, which seems to work nicely

So we end up changing one heuristic (sample sampling) with another (class sampling). What is the point then? :)

Regarding adding and/or removing classes - lets consider local patch descriptors for image matching. As you have almost infinite number of 3d points in the world, there are literally infinite number of the classes. The PhotoTour datasets contains only 3 scenes, 150k classes each. Recent Image Matching challenge proposes 26 scenes, each on containing 94k to 500k classes. If we sum it together, it is 200k * 26 = 5M classes. I can continue :)

> I am more troubled by your second point. What we wanted to show is that you can obtain competitive performance by completely discarding these complex sampling strategies. 

But at what cost? Softmax bottleneck is a real problem. I have recently tried CosFace (softmax on cosine dist + margin) loss for the PhotoTour dataset and it is 6 times slower than pairwise losses: it takes 120 min for 10M samples, while triplet loss takes 22 min for 10M samples.  I can hardly imagine anything more practical.

I am not against softmax loss at all - I am using it on daily basis, but I would like to point some problems with it. 

P.S. If you train model with CE on Liberty - please, let me know, I would be really surprised if you manage to have similar performance. I don't say that it is not possible, but I would be surprised. 
Btw, I have a starter pack for the problem - https://github.com/ducha-aiki/hardnet-in-fastai2-and-kornia . Feel free to use it.

**Comment by ibayed** [comment]
Hi Ducha-aiki, 

Thanks a lot for bringing the case of very large number of classes to our attention. This is extremely appreciated, and definitely deserves a least a discussion in a revised version of our paper (perhaps even additional experiments). Clearly, there is a bottleneck for CE for a large number of classes. Intuitively, and given the fact that there is a pairwise loss underlying the cross-entropy (Eq. 10 in the paper), sampling the classes is closely related to sampling data pairs (only pairs of sampled classes will appear in the bound in Eq. 10, which amounts to some implicit sampling strategy for pairs of points). Which way is better in practice (sampling classes or data points) is a very interesting research question, which definitely deserve further investigation. Chances are that this is problem-dependent.  

Ismail
